_source/_posts/2023-01-23-keycloak-kubernetes-prod.md

```diff
@@ -0,0 +1,637 @@
+- **Keycloak hostname**: Keycloak usually runs in a private network but certain public-facing endpoints must be exposed to applications. The base URLs determine how tokens are issued and validated, how action links are created, and how the OpenID Connect Discovery Document `realms/{realm-name}/.well-known/openid-configuration` is created.
+- **Reverse proxy**: A reverse proxy/load balancer component is recommended for a production environment, unifying access to the network. Keycloak supports multiple proxy modes. The `edge` mode allows HTTP communication between the proxy and Keycloak, and the proxy keeps a secure connection with clients.
+- **Production grade database**: The database plays a crucial role in the performance and Keycloak supports several production-grade databases, including PostgreSQL.
+- **High Availability**: Choose multi-mode clustered deployment. In production mode, distributed caching of realm and session data is enabled and all nodes in the network are discovered.
+
+## Deploy microservices and Keycloak to Google Cloud
+
+In this walkthrough, you will build a microservices architecture example from JHipster, consisting of a `gateway` application based on Spring Gateway and two reactive microservices `blog` and `store`. The `gateway` will act as the entrance to your microservices, providing HTTP routing and load balancing, quality of service, security and API documentation for all microservices. With the help of the JHipster Kubernetes sub-generator, you can deploy the application and its required services (Consul for service discovery, and Keycloak for OpenID Connect) to Google Kubernetes Engine (GKE).
+
+### Build a microservices architecture
+
+Install JHipster, you can do the classical local installation with npm.
+
+```bash
+npm install -g generator-jhipster@7.9.3
+```
+If you'd rather use Yarn or Docker, follow the instructions at [jhipster.tech](https://www.jhipster.tech/installation/#local-installation-with-npm-recommended-for-normal-users).
+
+Generate the microservices architecture from a JDL (JHipster-specific domain language) descriptor file. Create a folder for the project and add the file `reactive-ms.jdl` with the following content:
+
+```text
+application {
+  config {
+    baseName gateway
+    reactive true
+    packageName com.okta.developer.gateway
+    applicationType gateway
+    authenticationType oauth2
+    buildTool gradle
+    clientFramework vue
+    prodDatabaseType postgresql
+    serviceDiscoveryType consul
+    testFrameworks [cypress]
+  }
+  entities Blog, Post, Tag, Product
+}
+
+application {
+  config {
+    baseName blog
+    reactive true
+    packageName com.okta.developer.blog
+    applicationType microservice
+    authenticationType oauth2
+    buildTool gradle
+    databaseType neo4j
+    devDatabaseType neo4j
+    prodDatabaseType neo4j
```

Comment:
In v8, this is invalid. You only need `databaseType`. I ran into this today. 

https://github.com/jhipster/jdl-samples/issues/51

Reply:
I ran into more JDL issues: ERROR! ERROR! In the relationship between Blog and User, User is not declared. If 'User' is a built-in entity declare like 'Blog to User with builtInEntity'.

Reply:
Yep. If you look at the bottom of this PR's changes, you'll see I had to do this too. https://github.com/jhipster/jdl-samples/pull/52/files

Reply:
Updated the JDL